### ISSUE TITLE
Low depth search leads to miscalculation on long mates.

### ISSUE THREAD

**Issue by rwbc** [TheBlackPlague/StockNemo#18]
Hi, I did some test games with current release version.

I should mention that this was played on an old quadcore from 2009 ;-)
(to possibly excuse the low depths in that fast games)

http://talkchess.com/forum3/viewtopic.php?f=6&t=80259

Here is its opponent tested in the CCRL Blitz rating list in > 1000 games
http://ccrl.chessdom.com/ccrl/404/cgi/engine_details.cgi?print=Details&each_game=1&eng=Galjoen%200.33%2064-bit#Galjoen_0_33_64-bit

```

[Event "Test 2022"]
[Site "RWBC-CAPPUCCINO Win7U64 Q8200 2.33Ghz"]
[Date "2022.07.07"]
[Round "8.1.2"]
[White "StockNemo_2005-64"]
[Black "Galjoen_033-64"]
[Result "1/2-1/2"]
[ECO "A07"]
[GameDuration "00:04:37"]
[GameEndTime "2022-07-07T13:14:49.248 Mitteleuropäische Sommerzeit"]
[GameStartTime "2022-07-07T13:10:11.812 Mitteleuropäische Sommerzeit"]
[Opening "Reti"]
[PlyCount "126"]
[TimeControl "90+1"]
[Variation "King's Indian attack (Barcza system)"]

1. Nf3 {book} d5 {book}
2. g3 {book} c6 {book}
3. c4 {book} Nf6 {book}
4. cxd5 {-0.05/8 4.56} cxd5 {+0.21/9 3.98}
5. Bg2 {-0.05/6 4.37} Nc6 {+0.34/9 3.87}
6. O-O {-0.02/6 4.20} e5 {+0.83/9 3.78}
7. d4 {-0.19/6 4.55} exd4 {+0.57/9 3.66}
8. Nxd4 {+0.22/7 4.37} Bc5 {+0.62/9 3.64}
9. Nxc6 {+0.16/7 4.20} bxc6 {+0.74/8 3.51}
10. Qc2 {+0.16/5 4.04} Qb6 {+0.39/9 3.46}
11. Nc3 {+0.33/6 3.89} Bd4 {+0.30/8 3.36}
12. Bf4 {+0.54/6 3.74} Ba6 {-0.31/8 3.28}
13. Na4 {+0.90/6 3.60} Qb5 {-0.36/8 3.20}
14. Rad1 {+0.97/7 3.47} Bb6 {-0.30/9 3.12}
15. Nxb6 {+0.77/6 3.36} axb6 {-0.27/9 3.03}
16. Rfe1 {+0.84/5 3.24} O-O {-0.18/9 2.99}
17. e3 {+0.69/5 3.12} Rfe8 {+0.07/9 2.95}
18. Bf1 {+0.59/5 3.06} Qc5 {+0.28/8 2.85}
19. Qxc5 {+0.15/8 2.91} bxc5 {+0.48/9 2.81}
20. Bg2 {+0.32/7 2.81} Bc8 {+0.54/9 2.79}
21. a3 {+0.18/6 2.73} Bf5 {+0.44/9 2.72}
22. h3 {+0.03/6 2.64} c4 {+0.35/9 2.64}
23. g4 {-0.02/6 2.56} Bc2 {+0.31/10 2.58}
24. Rc1 {+0.05/7 2.48} Bd3 {+0.40/9 2.54}
25. g5 {+0.06/7 2.41} Nd7 {+0.50/9 2.46}
26. Bd6 {+0.17/5 2.33} f6 {+0.75/10 2.44}
27. h4 {+0.13/7 2.27} Kf7 {+0.72/9 2.43}
28. Rcd1 {+0.28/5 2.21} Ke6 {+0.96/8 2.33}
29. Bb4 {+0.22/6 2.15} c5 {+1.12/9 2.31}
30. Bc3 {+0.40/7 2.08} Kd6 {+0.95/9 2.27}
31. h5 {+0.55/7 2.04} fxg5 {+1.25/9 2.21}
32. Bxg7 {+0.52/7 1.98} Kc6 {+1.24/9 2.21}
33. h6 {+0.47/6 1.93} g4 {+1.26/10 2.16}
34. Kh2 {+0.63/6 1.88} Be4 {+1.08/9 2.12}
35. Bxe4 {+0.68/7 1.84} Rxe4 {+1.23/9 2.08}
36. f3 {+0.53/6 1.81} gxf3 {+2.73/10 2.02}
37. Kg3 {-0.35/7 1.77} Rae8 {+2.50/10 2.04}
38. Kxf3 {-0.22/8 1.73} d4 {+2.42/9 2.00}
39. a4 {-0.64/8 1.68} Nb6 {+3.10/8 1.95}
40. a5 {-0.32/9 1.65} Na4 {+3.52/9 1.89}
41. Rd2 {-0.42/8 1.62} Rxe3+ {+3.32/8 1.85}
42. Rxe3 {-0.82/10 1.58} Rxe3+ {+3.08/8 1.86}
43. Kg4 {-0.83/9 1.55} Nxb2 {+3.02/9 1.80}
44. Rxb2 {+1.85/9 1.53} c3 {+2.64/9 1.80}
45. Ra2 {+1.32/8 1.50} Rd3 {+3.19/9 1.75}
46. a6 {+1.89/8 1.48} Rd2 {+0.67/9 1.76}
47. Ra1 {+4.72/10 1.46} Re2 {+0.56/9 1.74}
48. a7 {+4.62/10 1.43} Re8 {+0.80/9 1.74}
49. a8=Q+ {+4.84/9 1.42} Rxa8 {+2.02/11 1.71}
50. Rxa8 {+4.08/12 1.39} c2 {1.65}
51. Ra6+ {+4.08/10 1.38} Kd7 {+1.51/9 1.66}
52. Ra1 {+4.37/10 1.35} Kd6 {+0.28/8 1.58}
53. Kf3 {+4.42/10 1.34} Kd5 {-2.74/9 1.59}
54. Kf4 {+4.42/9 1.31} c4 {-1.73/7 1.53}
55. Be5 {+4.75/11 1.30} c3 {-2.45/8 1.55}
56. Rc1 {+6.36/13 1.28} d3 {-3.24/9 1.53}
57. Bxc3 {+6.80/12 1.28} Kc4 {-8.47/10 1.52}
58. Bf6 {+7.03/12 1.25} d2 {-4.90/9 1.47}
59. Rxc2+ {+7.03/14 1.25} Kd3 {-4.92/10 1.50}
60. Rc3+ {+7.03/13 1.23} Ke2 {-4.79/11 1.48}
61. Rc2 {+7.03/12 1.23} Kd3 {1.45}
62. Rc3+ {+7.03/14 1.21} Ke2 {1.44}
63. Rc2 {+7.03/13 1.20} Kd3 {-4.93/11 1.43}
{Draw by 3-fold repetition}
1/2-1/2

[Event "Test 2022"]
[Site "RWBC-CAPPUCCINO Win7U64 Q8200 2.33Ghz"]
[Date "2022.07.07"]
[Round "6.1.2"]
[White "StockNemo_2005-64"]
[Black "Galjoen_033-64"]
[Result "0-1"]
[ECO "B23"]
[GameDuration "00:01:04"]
[GameEndTime "2022-07-07T12:53:38.549 Mitteleuropäische Sommerzeit"]
[GameStartTime "2022-07-07T12:52:34.167 Mitteleuropäische Sommerzeit"]
[Opening "Sicilian"]
[PlyCount "26"]
[TimeControl "90+1"]
[Variation "Closed"]

1. e4 {book} c5 {book}
2. Nc3 {book} e6 {book}
3. Nf3 {book} a6 {book}
4. Bd3 {+0.49/6 4.56} Nc6 {-0.15/9 3.98}
5. O-O {+0.34/6 4.38} g5 {+0.30/9 3.88}
6. Be2 {+0.81/7 4.21} g4 {+0.43/9 3.77}
7. Ne1 {+0.40/8 4.56} h5 {+0.69/9 3.70}
8. f3 {+0.40/6 4.36} Bd6 {+0.30/9 3.56}
9. Nd3 {+0.40/7 4.20} Bxh2+ {+1.82/8 3.50}
10. Kxh2 {+3.30/7 4.05} Qh4+ {+M1/3 0.055}
11. Kg1 {-M7/10 3.88} g3
12. Re1 {-M7/8 3.74} Qh2+
13. Kf1 {-M3/8 3.60} Qh1# {Black mates}
0-1
```

**Comment by TheBlackPlague** [comment]
@rwbc Thank you for bringing this to my attention. 

Here are my thoughts:
- It seems only 20 games were run? Well, that's all I see on the talkchess post? 20 games are by no indication a good example of how good an engine is compared to another. Head-to-head rating tests are already bad enough. Did you by any chance run more games?
- For the quick mate, it seems this is because the engine wasn't able to reach a high enough depth to see it is a mate. This is due to the lack of pruning which will be fixed in later releases. As of now, StockNemo only has alpha-beta and null move pruning. It's able to reach the depths it can with the assistance of transposition tables. It is very lacking in that aspect.
- As far as the draw goes, I'll look into that, but it's entirely possible that it was one of those situations where any other move resulted in a negative evaluation so a three-fold repetition draw was well, impossible to avoid. Though I imagine its evaluation should've mentioned that. I'll look into this.

**Comment by rwbc** [comment]
> @rwbc Thank you for bringing this to my attention.
> 
> Here are my thoughts:
> 
>     * It seems only 20 games were run? Well, that's all I see on the talkchess post? 20 games are by no indication a good example of how good an engine is compared to another. Head-to-head rating tests are already bad enough. Did you by any chance run more games?

I know this of course, because I am testing chess engines since around 1999 ;)
This was just a small stability test for a new program and not for a rating of any kind
to provide informations about possible major bugs and oddities.
(stability, time management, mating behaviour, repetition behaviour, stalemate problems, wrong 50 moves
implementation, and so on)

I am interested in chess programs of all strength! and my post was in no way intended to belittle
your program, but just to help with describing some obvious things.

I am also the maintainer of the XB/UCI chronology BTW and StockNemo was very recently added there.
https://rwbc-chess.de/

> 
>     * For the quick mate, it seems this is because the engine wasn't able to reach a high enough depth to see it is a mate. This is due to the lack of pruning which will be fixed in later releases. As of now, StockNemo only has alpha-beta and null move pruning. It's able to reach the depths it can with the assistance of transposition tables. It is very lacking in that aspect.
> 
>     * As far as the draw goes, I'll look into that, but it's entirely possible that it was one of those situations where any other move resulted in a negative evaluation so a three-fold repetition draw was well, impossible to avoid. Though I imagine its evaluation should've mentioned that. I'll look into this.

Well I thought repetition draws were recently fixed at least with a huge evaluation plus?
I don't know if you looked at the game so far?
White simply could take the P with the R and nothing can stop it to win the resulting endgame
K+B+P vs. K+P, because White will get the last pawn of Black as well and simply promote
its own last pawn. Thus it should avoid the repetition because giving away the R should still give at least
a material advantage of +3.x at least (positionally +999 of course)

For the mate I guess g3! was pruned away initially after which White is out of moves.

BTW you should mention in the readme somewhere that it needs an argument to play in a GUI
`--uci=True`, which I only found out by searching through your repo.

Good luck with StockNemo!

**Comment by TheBlackPlague** [comment]
> > @rwbc Thank you for bringing this to my attention.
> > Here are my thoughts:
> > ```
> > * It seems only 20 games were run? Well, that's all I see on the talkchess post? 20 games are by no indication a good example of how good an engine is compared to another. Head-to-head rating tests are already bad enough. Did you by any chance run more games?
> > ```
> 
> I know this of course, because I am testing chess engines since around 1999 ;) This was just a small stability test for a new program and not for a rating of any kind (that's why I also mentioned the ridiculous high error bars) to provide informations about possible major bugs and oddities (this involves no adjudication!). (stability, time management, mating behaviour, repetition behaviour, stalemate problems, wrong 50 moves implementation, and so on)
> 
> I am interested in chess programs of all strength! and my post was in no way intended to belittle your program, but just to help with describing some obvious things.
> 
> I am also the maintainer of the XB/UCI chronology BTW and StockNemo was very recently added there. https://rwbc-chess.de/

Hello.

Thank you for adding StockNemo to XB/UCI chronology. I really appreciate that.

I honestly wasn't expecting many people to start testing out StockNemo so early. Especially since the project has only been 2 months in development. It's quite a surprise that it even got noticed. I will add the `--uci=True` to the README. Technically, you shouldn't require it, but the normal launch clears the console window, which messes with many GUI (like cutechess) for which the argument was added.

I wasn't offended in any way or felt that StockNemo was belittled. Apparently, there's someone on talkchess who thinks a small number of games is a reasonable estimation for loads of games. I was unaware of your experience and just wanted to make it clear. 

> Well I thought repetition draws were recently fixed at least with a huge evaluation plus?
> I don't know if you looked at the game so far?

Repetition draws should be fixed and noticeable by StockNemo. See: [Checking for Three-fold Repetition](https://github.com/TheBlackPlague/StockNemo/blob/27c2cdbc1f4f91bb4dcf8a38010007c243e05896/Backend/Engine/MoveSearch.cs#L172)

I will investigate this game further and see what's going on. Once again, thank you for testing StockNemo and this insight is most valuable. ^^

> For the mate I guess g3! was pruned away initially after which White is out of moves.

The mate checks happen before alpha-beta even kicks in so there shouldn't be any pruning. This is probably a case where it may be fixed with extension on checks & higher depth search. I'll have to see what's going on to be sure though.

**Comment by TheBlackPlague** [comment]
As mentioned to you, @rwbc , I've looked into fixing the issues you mentioned and it seems (per the tests I've done) I have in fact fixed the Three-Fold-Repetition issue.

The PGN you posted where previously StockNemo decided `63. Rc2`, StockNemo now gives `Re3` which is indeed the best move. 

I will look into the mate issue as well and get to you with an update.

As always, thank you for testing StockNemo, and please continue to do so. The new release coming out today is 3.0.0.1 and will have the mentioned fix as well as be way better than 2.0.0.5 (I'd say maybe 400 ELO or more). I hope you update your list with the tests done on 3.0.0.1.

**Comment by TheBlackPlague** [comment]
> May be even the 'quick mate issue' is already solved (I would not call it a long mate BTW, but at least
it seemed to be out of the horizon of 2.0.0.5 under the used condition), because not only 3.0.0.1
(probably also 3.0.0.0) searches quite deeper now even on my old hardware, but it also never selects
the move 8. f3 which weakens the K-side.

Well, I'll close this issue then. Should there be any more issues you notice or see, feel free to make another issue.

Thank you for testing and adding StockNemo to your list.
